### PR TITLE
Fix bugs with extending the session handler class

### PR DESCRIPTION
**Pull request by sergant210**
### What does it do?
1. Fixes the bugs from issue #15928.
2. Refactoring inititialization of the session handler.

### How to test
See the issue.

### Related issue(s)/PR(s)
#15928.

### CODE REVIEW

File: core/src/Revolution/modX.php [hunk 1]
```diff
@@ -2746,25 +2746,14 @@ protected function _initSession($options = null) {
         $contextKey= $this->context instanceof modContext ? $this->context->get('key') : null;
         if ($this->getOption('session_enabled', $options, true) || isset($_GET['preview'])) {
             if (!in_array($this->getSessionState(), [modX::SESSION_STATE_INITIALIZED, modX::SESSION_STATE_EXTERNAL, modX::SESSION_STATE_UNAVAILABLE], true)) {
-                $sh = false;
-                if ($sessionHandlerClass = $this->getOption('session_handler_class', $options)) {
-                    if ($shClass = $this->loadClass($sessionHandlerClass, '', false, true)) {
-                        if ($sh = new $shClass($this)) {
-                            session_set_save_handler(
```
Comment (Mark-H) [inline core/src/Revolution/modX.php:2753]: This line is removed but not added back anywhere that I can see - so it seems with this patch we're not overriding the session handler anymore

File: core/src/Revolution/modX.php [hunk 2]
```diff
@@ -2746,25 +2746,14 @@ protected function _initSession($options = null) {
...
+                    $ch = new $sessionHandlerClass($this);
+                    if ($ch instanceof \SessionHandlerInterface) {
+                        $this->services->add('session_handler', $ch);
                     }
                 }
-                if (
-                    (is_string($sessionHandlerClass) && !$sh instanceof $sessionHandlerClass) ||
-                    !is_string($sessionHandlerClass)
-                ) {
+                if (!$this->services->has('session_handler')) {
```
Comment (opengeek) [inline core/src/Revolution/modX.php:2757]: I'm not sure this is correct. You do not need a custom session handler to be able to set the session_save_path. In fact, this is meant to be used with the default PHP session handling to customize the location of the session files.

File: core/src/Revolution/modX.php [hunk 3]
```diff
@@ -2746,25 +2746,14 @@ protected function _initSession($options = null) {
...
-                        if ($sh = new $shClass($this)) {
-                            session_set_save_handler(
-                                [& $sh, 'open'],
-                                [& $sh, 'close'],
-                                [& $sh, 'read'],
-                                [& $sh, 'write'],
-                                [& $sh, 'destroy'],
-                                [& $sh, 'gc']
-                            );
-                        }
```
Comment (opengeek) [inline core/src/Revolution/modX.php:2761]: Where is the session handler being registered if this is removed?

File: core/src/Revolution/modX.php (hunk 1 above)
Comment (sergant210) [reply core/src/Revolution/modX.php:2753]: Look at line 2749.

Comment (Mark-H) [reply core/src/Revolution/modX.php:2753]: I'm not seeing another instance of `session_set_save_handler` anywhere in this diff.

Comment (sergant210) [reply core/src/Revolution/modX.php:2753]: Oops, I forgot to move it from my test site. Thanks.

File: core/src/Revolution/modX.php (hunk 3 above)
Comment (sergant210) [reply core/src/Revolution/modX.php:2761]: It's my fault. I forgot to add this `session_set_save_handler($sh);`.

File: core/src/Revolution/modX.php (hunk 2 above)
Comment (sergant210) [reply core/src/Revolution/modX.php:2757]: Can you show me an example to make this condition true?
```
if (
    (is_string($sessionHandlerClass) && !$sh instanceof $sessionHandlerClass) ||
     !is_string($sessionHandlerClass)
)
```

Comment (opengeek) [reply core/src/Revolution/modX.php:2757]: If it is not set to a modSessionHandler class or it is empty so it uses what is defined in PHP's ini.

Comment (sergant210) [reply core/src/Revolution/modX.php:2757]: 1. There is a deprecated loadClass method.
2. `$sh` object will always be an instance of `$sessionHandlerClass`, so this condition will be true only if `$sessionHandlerClass` is not a string or "".
3. I've moved this condition to line 2750.

File: core/src/Revolution/modX.php (hunk 1 above)
Comment (sergant210) [reply core/src/Revolution/modX.php:2753]: Fixed.

**Comment by JoshuaLuckers** [comment]
@cla-bot check

**Comment by rthrash** [comment]
This pull request has been mentioned on **MODX Community**. There might be relevant details there:

https://community.modx.com/t/error-flush-sessions-not-supported-when-trying-to-logout-all-users-in-modx-3/6361/2

**Comment by rthrash** [comment]
This pull request has been mentioned on **MODX Community**. There might be relevant details there:

https://community.modx.com/t/error-flush-sessions-not-supported-when-trying-to-logout-all-users-in-modx-3/6361/7

**Comment by Mark-H** [comment]
Closing in favour of #16522 - please see my explanation there.

**Pull request closed without merging**